httpserver/swarm.cpp

```diff
@@ -34,80 +31,110 @@ SwarmEvents Swarm::update_swarms(const all_swarms_t& swarms) {
         }
+    }
+
+    /// See if there are any new swarms
 
-        /// don't bother checking the rest
-        if (!events.decommissioned) {
+    for (const auto& swarm_info : swarms) {
 
-            /// See if anyone joined our swarm
-            for (const auto& sn : our_swarm_snodes) {
+        const bool found = std::any_of(
+            all_cur_swarms_.begin(), all_cur_swarms_.end(),
+            [&swarm_info](const SwarmInfo& cur_swarm_info) {
+                return cur_swarm_info.swarm_id == swarm_info.swarm_id;
+            });
 
-                const auto it =
-                    std::find(swarm_peers_.begin(), swarm_peers_.end(), sn);
+        if (!found) {
+            events.new_swarms.push_back(swarm_info.swarm_id);
+        }
+    }
 
-                if (it == swarm_peers_.end() && sn != our_address_) {
-                    LOKI_LOG(info, "EVENT: detected new SN: {}", sn);
-                    events.new_snodes.push_back(sn);
-                }
-            }
+    /// NOTE: need to be careful and make sure we don't miss any
+    /// swarm update (e.g. if we don't update frequently enough)
 
-            /// See if there are any new swarms
+    return events;
+}
 
-            for (const auto& swarm_info : swarms) {
+void Swarm::set_swarm_id(swarm_id_t sid) {
 
-                const bool found = std::any_of(
-                    all_cur_swarms_.begin(), all_cur_swarms_.end(),
-                    [&swarm_info](const SwarmInfo& cur_swarm_info) {
-                        return cur_swarm_info.swarm_id == swarm_info.swarm_id;
-                    });
+    if (sid == INVALID_SWARM_ID) {
+        LOKI_LOG(warn, "We are not currently an active Service Node");
+    } else {
 
-                if (!found) {
-                    LOKI_LOG(info, "EVENT: detected a new swarm: {}",
-                             swarm_info.swarm_id);
-                    events.new_swarms.push_back(swarm_info.swarm_id);
-                }
-            }
+        if (cur_swarm_id_ == INVALID_SWARM_ID) {
+            LOKI_LOG(info, "EVENT: started SN in swarm: {}", sid);
+        } else if (cur_swarm_id_ != sid) {
+            LOKI_LOG(info, "EVENT: got moved into a new swarm: {}", sid);
         }
     }
 
-    /// NOTE: need to be careful and make sure we don't miss any
-    /// swarm update (e.g. if we don't update frequently enough)
+    cur_swarm_id_ = sid;
+}
+
+void Swarm::update_state(const all_swarms_t& swarms, const SwarmEvents& events) {
+
+    if (events.decommissioned) {
+        LOKI_LOG(info, "EVENT: our old swarm got DISSOLVED!");
+    }
+
+    for (const sn_record_t& sn : events.new_snodes) {
+        LOKI_LOG(info, "EVENT: detected new SN: {}", sn);
+    }
+
+    for (swarm_id_t swarm : events.new_swarms) {
+        LOKI_LOG(info, "EVENT: detected a new swarm: {}", swarm);
+    }
 
-    cur_swarm_id_ = our_swarm_id;
     all_cur_swarms_ = swarms;
 
+    const auto& members = events.our_swarm_members;
+
+    if (members.size() <= 1)
```

Comment:
Maybe this should check `members.size() == 0` for integration tests/sanity?

Reply:
updated